### PR TITLE
New domain join and add admin user artifacts

### PR DESCRIPTION
**Pull request by mmydland**
- Added a new version of the domain join artifact based on custom script extension
    - Adds machine to the domain
    - Adds a user to the admin group of that domain
- Added an artifact to add a domain user to the local admin group of the target VM

### CODE REVIEW

**Comment by leovms** [comment]
In future, the preference is to provide separate check-ins for each artifact.

File: Artifacts/windows-add-local-admin/artifact.ps1 [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    {
+        $ADResolved = ($Username -split '@')[0]
+        $DomainResolved = ($Username -split '@')[1]    
+        $Username = 'WinNT://', "$DomainResolved", '/', $ADResolved -join ''
+    }
+    else
+    {
+        $ADResolved = ($Username -split '\\')[1]
+        $DomainResolved = ($Username -split '\\')[0]
+        $Username = 'WinNT://', $DomainResolved, '/', $ADResolved -join ''
```
Comment (leovms) [inline Artifacts/windows-add-local-admin/artifact.ps1]: Ditto on the syntax. In fact, you can remove this line from both the `if / else` condition and just return it:
`return "WinNT://$DomainResolved/$ADResolved"`. The `if / else` condition is only required to get the correct parts. Once we have them we just need to return the AD user name string.

File: Artifacts/windows-add-local-admin/artifact.ps1 [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+function Get-LocalAdminGroupName ()
+{
+    #well known SID for the Administrators group
+    ([wmi]"Win32_SID.SID='S-1-5-32-544'").AccountName
+}
+
+#############################
+# Function to find the members of a local machine group, usually administrators.
+# Works around the cahnge in call pattern InvokeMember
+function Get-LocalGroupMembersPs3to5 ([Parameter(Mandatory = $true)] [string] $GroupName)
```
Comment (leovms) [inline Artifacts/windows-add-local-admin/artifact.ps1]: Move all parameters to a `param` section inside the function, for all functions. For example,
```
function Get-LocalGroupMembersPs3to5
{
    param(
        ([Parameter(Mandatory = $true)]
        [string] $GroupName
    )
    ...
}
```

File: Artifacts/windows-add-local-admin/artifact.ps1 [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+            if ($path -like "*/$env:COMPUTERNAME/*")
+            {
+                $Type = "Local"
+            }
+           
+            $Details += New-Object PSObject -Property @{"Account" = $name; "Type" = $type; }
+        }
+    }
+    Write-Output "Members of $GroupName are:"
+    $Details
```
Comment (leovms) [inline Artifacts/windows-add-local-admin/artifact.ps1]: `return $Details`

File: Artifacts/windows-add-local-admin/artifact.ps1 [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+            $Type = "Domain"
+            if ($path -like "*/$env:COMPUTERNAME/*")
+            {
+                $Type = "Local"
+            }
+           
+            $Details += New-Object PSObject -Property @{"Account" = $name; "Type" = $type; }
+        }
+    }
+    Write-Output "Members of $GroupName are:"
```
Comment (leovms) [inline Artifacts/windows-add-local-admin/artifact.ps1]: Remove from here and move to main body before function call.

File: Artifacts/windows-add-local-admin/artifact.ps1 [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    {
+        Write-Error "Result: failed to add $username to $GroupName"
+    }
+}
+
+##############################
+# Main function
+$adminGroupName = Get-LocalAdminGroupName
+Add-UserToLocalGroup -Username $DomainJoinUsername -GroupName $adminGroupName
+Get-LocalGroupMembers $adminGroupName
```
Comment (leovms) [inline Artifacts/windows-add-local-admin/artifact.ps1]: Add `Write-Output "Members of $GroupName are:"` a line before calling `Get-LocalGroupMembers $adminGroupName`.

File: Artifacts/windows-add-local-admin/artifact.ps1 [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+            $Type = "Domain"
+            if ($path -like "*/$env:COMPUTERNAME/*")
+            {
+                $Type = "Local"
+            }
+           
+            $Details += New-Object PSObject -Property @{"Group" = $GroupName; "Account" = $name; "Type" = $type; }
+        }
+    }
+    Write-Output "Members of $GroupName are:"
```
Comment (leovms) [inline Artifacts/windows-add-local-admin/artifact.ps1]: Remove from here and move to main body before calling function.

File: Artifacts/windows-add-local-admin/artifact.ps1 [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+    $Details.ForEach( {$_}) | Format-Table -Property Account, Type -AutoSize
```
Comment (leovms) [inline Artifacts/windows-add-local-admin/artifact.ps1]: This function should only return the data and the formatting should be done in the main body where the call happens. Specially true as we have 2 functions, doing very similar work, but now formatting the output differently (i.e. `Get-LocalGroupMembersPs3to5`).

File: Artifacts/windows-domain-join-new/README.md [hunk 8]
```diff
@@ -0,0 +1 @@
+This artifact joins the virtual machine to the corporate domain.  You must provide your user account which will be added to the administrator group on the virtual machine.  After execution of the artifact it will be rebooted automatically and then you will be able to login with your user account directly on the virtual machine.
```
Comment (leovms) [inline Artifacts/windows-domain-join-new/README.md]: Ditto, do we need the file? If so, is the structure good enough?

File: Artifacts/windows-domain-join-new/artifact-funcs.ps1 [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+    }
+    
+    # IMPORTANT NOTE: Throwing a terminating error (using $ErrorActionPreference = "Stop") still
+    # returns exit code zero from the PowerShell script when using -File. The workaround is to
+    # NOT use -File when calling this script and leverage the try-catch-finally block and return
+    # a non-zero exit code from the catch block.
+    exit -1
+}
+
+function DownloadToFilePath ($downloadUrl, $targetFile)
```
Comment (leovms) [inline Artifacts/windows-domain-join-new/artifact-funcs.ps1]: Remove.

File: Artifacts/windows-domain-join-new/artifact-funcs.ps1 [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+
+    } while ($downloadAttempts -lt 5)
+
+    if($downloadAttempts -eq 5)
+    {
+        Write-Error "Download of $downloadUrl failed repeatedly. Giving up."
+    }
+}
+
+function InstallChocoPackages ($packageList)
```
Comment (leovms) [inline Artifacts/windows-domain-join-new/artifact-funcs.ps1]: Remove.

File: Artifacts/windows-domain-join-new/artifact-funcs.ps1 [hunk 11]
```diff
@@ -0,0 +1,76 @@
+﻿function Handle-LastError
```
Comment (leovms) [inline Artifacts/windows-domain-join-new/artifact-funcs.ps1]: See comment for other artifact.

File: Artifacts/windows-domain-join-new/artifact.ps1 [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+{
+    Write-Error "The current version of PowerShell is $($PSVersionTable.PSVersion.Major). Prior to running this artifact, ensure you have PowerShell 3 or higher installed."
+}
+
+else
+{
+    $adminGroupName = Get-LocalAdminGroupName
+    $securePass = ConvertTo-SecureString $DomainAdminPassword -AsPlainText -Force
+    Add-VmToDomain -VmName $env:COMPUTERNAME -DomainName $DomainToJoin -JoinUser $DomainAdminUsername -JoinPassword $securePass -OU $OUPath
+    Add-UserToLocalGroup -Username $DomainJoinUsername -GroupName $adminGroupName
```
Comment (leovms) [inline Artifacts/windows-domain-join-new/artifact.ps1]: Let's not mix 2 artifacts' functionality into 1. I won't review the part dealing with adding the user to the local Administrators group, as all the comments already done in that artifact apply. Let the caller chain these 2 artifacts, or make a single artifact and get rid of the other one. No real advantage duplicating the code here, I don't think.

File: Artifacts/windows-add-local-admin/Artifactfile.json [hunk 13]
```diff
@@ -0,0 +1,21 @@
+{
+  "$schema": "https://raw.githubusercontent.com/Azure/azure-devtestlab/master/schemas/2015-01-01/dtlArtifacts.json",
+  "title": "Add user to Administrators Group",
+  "description": "This artifact adds the given Active Directory account to the Local Administrator group of the target virtual machine.  The virtual machine must be joined to the domain before this artifact is run.",
```
Comment (mmydland) [inline Artifacts/windows-add-local-admin/Artifactfile.json]: done

File: Artifacts/windows-add-local-admin/Artifactfile.json [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+  "tags": [
+    "Windows"
+  ],
+  "iconUri": "https://raw.githubusercontent.com/Azure/azure-devtestlab/master/Artifacts/windows-domain-join/domainJoinArtifact.png",
+  "targetOsType": "Windows",
+  "parameters": {
+    "domainJoinUsername": {
+      "type": "string",
+      "displayName": "Domain user account",
+      "description": "The user account (e.g. CORP\\user or user@corp.contoso.com) that will be added to the administrators group on the computer."
```
Comment (mmydland) [inline Artifacts/windows-add-local-admin/Artifactfile.json]: done

File: Artifacts/windows-add-local-admin/README.md [hunk 15]
```diff
@@ -0,0 +1 @@
+This artifact will add the given Active Directory (i.e. domain) user to the local administrators group on the target virtual machine.
```
Comment (mmydland) [inline Artifacts/windows-add-local-admin/README.md]: agreed.  removed it for now

File: Artifacts/windows-add-local-admin/artifact-funcs.ps1 [hunk 16]
```diff
@@ -0,0 +1,76 @@
+﻿function Handle-LastError
```
Comment (mmydland) [inline Artifacts/windows-add-local-admin/artifact-funcs.ps1]: Once I made the change below, the functions are identical.

File: Artifacts/windows-add-local-admin/artifact-funcs.ps1 [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+{
+    [CmdletBinding()]
+    param(
+    )
+
+    $message = $error[0].Exception.Message
+    if ($message)
+    {
+        Write-Host -Object "ERROR: $message" -ForegroundColor Red
+        Write-Host -Object $error -ForegroundColor Red
```
Comment (mmydland) [inline Artifacts/windows-add-local-admin/artifact-funcs.ps1]: done

File: Artifacts/windows-add-local-admin/artifact-funcs.ps1 [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+    }
+    
+    # IMPORTANT NOTE: Throwing a terminating error (using $ErrorActionPreference = "Stop") still
+    # returns exit code zero from the PowerShell script when using -File. The workaround is to
+    # NOT use -File when calling this script and leverage the try-catch-finally block and return
+    # a non-zero exit code from the catch block.
+    exit -1
+}
+
+function DownloadToFilePath ($downloadUrl, $targetFile)
```
Comment (mmydland) [inline Artifacts/windows-add-local-admin/artifact-funcs.ps1]: removed

File: Artifacts/windows-add-local-admin/artifact.ps1 [hunk 19]
```diff
@@ -0,0 +1,137 @@
...
+    [Parameter(Mandatory = $true)]
+    [string] $DomainJoinUsername
+)
+
+##############################
+# Function returns the localized group name for the local administrator group
+function Get-LocalAdminGroupName ()
+{
+    #well known SID for the Administrators group
+    ([wmi]"Win32_SID.SID='S-1-5-32-544'").AccountName
```
Comment (mmydland) [inline Artifacts/windows-add-local-admin/artifact.ps1]: done

File: Artifacts/windows-add-local-admin/artifact.ps1 (hunk 2 above)
Comment (mmydland) [inline Artifacts/windows-add-local-admin/artifact.ps1]: done

File: Artifacts/windows-add-local-admin/artifact.ps1 (hunk 4 above)
Comment (mmydland) [inline Artifacts/windows-add-local-admin/artifact.ps1]: done

File: Artifacts/windows-add-local-admin/artifact.ps1 (hunk 6 above)
Comment (mmydland) [inline Artifacts/windows-add-local-admin/artifact.ps1]: done

File: Artifacts/windows-add-local-admin/artifact.ps1 (hunk 7 above)
Comment (mmydland) [inline Artifacts/windows-add-local-admin/artifact.ps1]: done.

File: Artifacts/windows-add-local-admin/artifact.ps1 [hunk 20]
```diff
@@ -0,0 +1,137 @@
...
+    $Details.ForEach( {$_}) | Format-Table -Property Account, Type -AutoSize
+}
+    
+##############################
+# Function to get the AD User string in format "WinNT://<domain>/<username>"
+function Get-AdUsernameString ([Parameter(Mandatory = $true)] [string] $UserName)
+{
+    if ([string]::IsNullOrEmpty($Username)) 
+    { 
+        Write-Error "Username not provided"
```
Comment (mmydland) [inline Artifacts/windows-add-local-admin/artifact.ps1]: done

File: Artifacts/windows-add-local-admin/artifact.ps1 (hunk 1 above)
Comment (mmydland) [inline Artifacts/windows-add-local-admin/artifact.ps1]: done

File: Artifacts/windows-add-local-admin/artifact.ps1 (hunk 5 above)
Comment (mmydland) [inline Artifacts/windows-add-local-admin/artifact.ps1]: done

File: Artifacts/windows-domain-join-new/README.md (hunk 8 above)
Comment (mmydland) [inline Artifacts/windows-domain-join-new/README.md]: removed

File: Artifacts/windows-domain-join-new/artifact-funcs.ps1 (hunk 11 above)
Comment (mmydland) [inline Artifacts/windows-domain-join-new/artifact-funcs.ps1]: done

File: Artifacts/windows-domain-join-new/artifact.ps1 (hunk 12 above)
Comment (mmydland) [inline Artifacts/windows-domain-join-new/artifact.ps1]: yep, agreed.  Pulled the admin join stuff our.

**Pull request merged**